pkg/observability/opentelemetry_test.go

```diff
@@ -0,0 +1,239 @@
+	const errorActionMetricsTotal = 97
+
+	metrics := getMetrics(t)
+
+	errorEventLables := []attribute.KeyValue{labelEventErrorWhereKey.String(mockErrorEvent)}
+	successActionLables := []attribute.KeyValue{labelNodeActionKey.String(mockAction), labelNodeStatusKey.String(successStatus)}
+	errorActionLables := []attribute.KeyValue{labelNodeActionKey.String(mockAction), labelNodeStatusKey.String(errorStatus)}
+
+	for i := 0; i < errorEventMetricsTotal; i++ {
+		metrics.errorEventsCounter.Add(context.Background(), 1, api.WithAttributes(errorEventLables...))
+	}
+	for i := 0; i < successActionMetricsTotal; i++ {
+		metrics.actionsCounterV2.Add(context.Background(), 1, api.WithAttributes(successActionLables...))
+	}
+	for i := 0; i < errorActionMetricsTotal; i++ {
+		metrics.actionsCounterV2.Add(context.Background(), 1, api.WithAttributes(errorActionLables...))
+	}
+
+	rr := mockMetricsRequest()
+
+	validateStatus(t, rr)
+
+	metricsMap := getMetricsMap(rr.Body.String())
+
+	validateEventErrorTotal(t, metricsMap, errorEventMetricsTotal)
+	validateActionTotalV2(t, metricsMap, successActionMetricsTotal, successStatus)
+	validateActionTotalV2(t, metricsMap, errorActionMetricsTotal, errorStatus)
+}
+
+func TestServeMetrics(t *testing.T) {
+	server := serveMetrics(mockDefaultPort)
+
+	defer func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		if err := server.Shutdown(ctx); err != nil {
+			t.Errorf("failed to shutdown server: %v", err)
+		}
+	}()
+
+	time.Sleep(100 * time.Millisecond)
+
+	conn, err := net.DialTimeout("tcp", fmt.Sprintf("localhost:%d", mockDefaultPort), time.Second)
+	if err != nil {
+		t.Errorf("server not listening on port %d: %v", mockDefaultPort, err)
+	}
+	conn.Close()
+
+	conn, err = net.DialTimeout("tcp", fmt.Sprintf("localhost:%d", mockClosedPort), time.Second)
+	if err == nil {
+		conn.Close()
+		t.Errorf("server should not listening on port %d: %v", mockClosedPort, err)
+	}
+}
+
+func getMetrics(t *testing.T) *Metrics {
+	exporter, err := prometheus.New()
+	if err != nil {
+		t.Errorf("failed to create Prometheus exporter: %v", err)
+	}
+	provider := metric.NewMeterProvider(metric.WithReader(exporter))
+	metrics, err := registerMetricsWith(provider)
+	if err != nil {
+		t.Errorf("failed to register metrics with Prometheus provider: %v", err)
+	}
+	metrics.enabled = true
+
+	t.Cleanup(func() {
+		if provider != nil {
+			if err := provider.Shutdown(context.Background()); err != nil {
+				t.Errorf("failed to shutdown provider: %v", err)
+			}
+		}
+	})
+
+	return &metrics
+}
+
+func mockMetricsRequest() *httptest.ResponseRecorder {
+	handler := promhttp.Handler()
+	req := httptest.NewRequest("GET", metricsEndpoint, nil)
+	rr := httptest.NewRecorder()
+	handler.ServeHTTP(rr, req)
+	return rr
+}
+
+func validateStatus(t *testing.T, rr *httptest.ResponseRecorder) {
+	status := rr.Code
+	h.Equals(t, http.StatusOK, status)
+}
+
+func getMetricsMap(body string) map[string]string {
```

Comment:
Could you add a comment here indicating an example of what the body might look like? Is there only a specific format it accepts? Just so it's easier to understand what this does?

Reply:
Fixed by adding doc comment for example response body.